gateway/src/apicast/policy/3scale_batcher/3scale_batcher.lua

```diff
@@ -0,0 +1,160 @@
+local function format_usage(usage)
+  local res = {}
+
+  local usage_metrics = usage.metrics
+  local usage_deltas = usage.deltas
+
+  for _, metric in ipairs(usage_metrics) do
+    local delta = usage_deltas[metric]
+    res['usage[' .. metric .. ']'] = delta
+  end
+
+  return res
+end
+
+local function set_flags_to_avoid_auths_in_apicast(context)
+  context.skip_apicast_access = true
+  context.skip_apicast_post_action = true
+end
+
+local function report(_, service_id, backend, reports_batcher)
+  local reports = reports_batcher:get_all(service_id)
+  reporter.report(reports, service_id, backend, reports_batcher)
+end
+
+-- This starts a timer on each worker.
+-- Starting a timer on each worker means that there will be more calls to
+-- 3scale backend, and the config param 'batch_report_seconds' becomes
+-- more confusing because the reporting frequency will be affected by the
+-- number of APIcast workers.
+-- If we started a timer just on one of the workers, it could die, and then,
+-- there would not be any reporting.
+local function ensure_report_timer_on(self, service_id, backend)
+  local check_timer = self.semaphore_report_timer:wait(0)
+
+  if check_timer then
+    if not self.report_timer_on then
+      ngx.timer.every(self.batch_reports_seconds, report,
+        service_id, backend, self.reports_batcher)
+
+      self.report_timer_on = true
+    end
+
+    self.semaphore_report_timer:post()
+  end
+end
+
+local function rejection_reason_from_headers(response_headers)
+  return response_headers and response_headers['3scale-rejection-reason']
+end
+
+local function error(service, rejection_reason)
+  if rejection_reason == 'limits_exceeded' then
+    return errors.limits_exceeded(service)
+  else
+    return errors.authorization_failed(service)
+  end
+end
+
+function _M:access(context)
+  local backend = backend_client:new(context.service, http_ng_resty)
+  local usage = context.usage
+  local service = context.service
+  local service_id = service.id
+  local credentials = context.credentials
+
+  ensure_report_timer_on(self, service_id, backend)
+
+  local cached_auth = self.auths_cache:get(service_id, credentials, usage)
+
+  if not cached_auth then
+    local formatted_usage = format_usage(usage)
+    local backend_res = backend:authorize(formatted_usage, credentials)
+    local backend_status = backend_res.status
+
+    if backend_status == 200 then
+      self.auths_cache:set(service_id, credentials, usage, 200)
+      self.reports_batcher:add(service_id, credentials, usage)
+    elseif backend_status >= 400 and backend_status < 500 then
+      local rejection_reason = rejection_reason_from_headers(backend_res.headers)
+      self.auths_cache:set(service_id, credentials, usage, backend_status, rejection_reason)
+      return error(service, rejection_reason)
+    else
+      return error(service)
+    end
+  else
+    if cached_auth.status == 200 then
+      self.reports_batcher:add(service_id, credentials, usage)
```

Comment:
This could go to `:log` phase so client does not see the impact.

Reply:
Good idea :+1: 
Why `log` instead of `post_action` ?

Reply:
We can't do that @mikz 

`resty.lock.lock()` calls `ngx.sleep()` which cannot be run in the log phase:
https://github.com/openresty/lua-resty-lock/blob/master/lib/resty/lock.lua#L151
https://github.com/openresty/lua-nginx-module#ngxsleep

I realized because I saw errors like this one in the logs:
```
[error] 7127#7127: *150 failed to run log_by_lua*: /usr/local/openresty/lualib/resty/lock.lua:151: API disabled in the context of log_by_lua*
```

Reply:
It would be possible to do that by creating a timer.at(0) to run in the background, but that we can explore later.